new ZI driver files from DiCarloLab-Delft/PycQED_py3

chellings opened the pull request:
new ZI driver files from DiCarloLab-Delft/PycQED_py3

Changes from Delft and various PRs (https://github.com/DiCarloLab-Delft/PycQED_py3/pulls?q=is%3Apr+author%3Anielshaandbaek) by Niels H., including compatibility with LabOne 20.08.

The code in this pull request is a merge of our changes and the changes in Delft since the last big update (fall 2019). I have done this merge mostly based on my knowledge of git, not based on my knowledge of the code ;-). So it is necessary that people who know these drivers better than me have a look and check whether all changes from Delft make sense for us. Inviting @antsr and @MichKe95 to do this :-). Thank you!

I create this as a draft pull request to make it easier for you to see the changes. We should turn it into a real pull request after some tests on a setup.

To see how I merged our code with the code from DiCarloLab-Delft/PycQED_py3, check out the branch devel/ZI_driver_merge. The result of this merge is added here as commit 2a61e58. Retrospectively, it looks easy, but it took me quite a while to figure out how I can merge something between these completely diverged repositories. Keeping and maintaining the branch devel/ZI_driver_merge will facilitate future driver updates.

FYI @nathlacroix @stephlazar

chellings committed:
ZI_HDAWG_core: revert code formatting to simplify merging

chellings committed:
ZI_base_instrument: restructured PQSC handling in start()

chellings committed:
new ZI driver files from  DiCarloLab-Delft/PycQED_py3
Changes from Delft and various PRs from Niels H., including compatibility with LabOne 20.08.

chellings committed:
utilities/general: added check_keyboard_interrupt (required by new zi driver files)

chellings committed:
ZI_base_instrument: log an info instead of a warning if an awg is unused

MichKe95 commented on pycqed/instrument_drivers/physical_instruments/ZurichInstruments/UHFQuantumController.py:
```diff
@@ -199,24 +214,89 @@ def __init__(self,
+            if params_to_exclude and name in params_to_exclude:
+                pass
+            elif params_to_skip_update and name in params_to_skip_update:
+                update_par = False
```
This parameter does not seem to have any effect.

chellings replied:
If it does not hurt, let's just keep it :-). We need to stay in sync with the DCL version as closely as possible in order to enable further driver updates in the future. I would say:
- If we spot things in their code that are definitely wrong and have a negative effect, we can create pull requests in their git.
- If we spot things that we definitely need differently, but that look fine for them, we have to change them in our fork (and live with the trouble in future merges).
- If we spot things that do not really hurt, let's be pragmatic and just leave them as they are for the sake of allowing easy merging (same for code formatting!).

MichKe95 replied:
Agree!

MichKe95 commented on pycqed/instrument_drivers/physical_instruments/ZurichInstruments/UHFQuantumController.py:
```diff
@@ -199,24 +214,89 @@ def __init__(self,
+
+    def snapshot_base(self, update: bool=False,
+                      params_to_skip_update =None,
+                      params_to_exclude = None ):
```
- Not mentioned in the doctoring. 
- Also it is not clear to me, how the effect of this parameter differs from params_to_skip.

MichKe95 commented on pycqed/instrument_drivers/physical_instruments/ZurichInstruments/UHFQuantumController.py:
```diff
@@ -199,24 +214,89 @@ def __init__(self,
 
     ##########################################################################
-    # Private methods
+    # Overriding Qcodes InstrumentBase methods
```
Would be nice to know why the base function was overwritten. (major change seems to be that this function only logs to .info and not to .warning as well, as is the case for the base function.)

MichKe95 commented on pycqed/instrument_drivers/physical_instruments/ZurichInstruments/UHFQuantumController.py:
```diff
@@ -734,14 +920,48 @@ def acquisition_poll(self, samples, arm=True,
 
         return data
 
+    def acquisition_get(self, samples, arm=True,
```
This seems to be an additional function for retrieving data from the UHFQA (the other being acquisition_poll). It is not entirely clear to me in how they differ and in which instances one or the other routine should be called (the default one still being acquisition_poll, see self.acquisition()).

Note: Currently we implement the polling in the UHFQC_Base class. As it accesses private attributes, we might consider using the routines provided in the ZI driver.

chellings commented:
Short update: successfully tested on a virtual setup and with real instruments on XLD (still old firmware, and no quantum device connected).

MichKe95 reviewed (approved):
- it is difficult to assure on such a low level, whether any of our higher-level code will be broken by the changes, but
- after a quick scan it seems that most changes are to routines only relevant for DiCarlo Lab, are code formatting or added logging statements
- for some of the remaining points see my comments above (nothing critical though)

I think the next step would be to test the software on actual hardware (best case a test setup/ not at XLD).

chellings committed:
Pulsar: remove prefetching commands for LabOne 20.11 compatibility

chellings committed:
pulsar: added option to append a playZero command to all ZI HDAWG playback strings
This can be activated by setting the property append_zeros of pulsar to the desired number of zero samples.

antsr commented:
These changes have been in use now for quite some time on both of our pycqed setups. I believe we should also add the removal of the prefetch statements from LabOne 20.11 to this PR and then merge it.

antsr reviewed (approved):
Ready to merge by me

antsr committed:
Pulsar: upgrade append_zeros to a qcodes parameter

chellings committed:
Merge branch 'qudev_master' into enh/zi_driver_update
